pkg/sriov/sriov.go

```diff
@@ -0,0 +1,113 @@
+//
+// Use and distribution licensed under the Apache license version 2.
+//
+// See the COPYING file in the root project directory for full text.
+//
+
+package sriov
+
+import (
+	"fmt"
+
+	"github.com/jaypipes/ghw/pkg/context"
+	"github.com/jaypipes/ghw/pkg/marshal"
+	"github.com/jaypipes/ghw/pkg/option"
+	"github.com/jaypipes/ghw/pkg/pci"
+	pciaddr "github.com/jaypipes/ghw/pkg/pci/address"
+)
+
+type Device struct {
+	Interfaces []string `json:"interfaces"`
+	// the PCI address where the SRIOV instance can be found
+	Address *pciaddr.Address `json:"address"`
+	PCI     *pci.Device      `json:"pci"`
+}
+
+func (d Device) ToString(devType string) string {
+	deviceStr := d.Address.String()
+	nodeStr := ""
+	if d.PCI != nil {
+		deviceStr = d.PCI.String()
+		if d.PCI.Node != nil {
+			nodeStr = fmt.Sprintf(" [affined to NUMA node %d]", d.PCI.Node.ID)
+		}
+	}
+	return fmt.Sprintf("%s function %s@%s", devType, nodeStr, deviceStr)
+}
+
+type PhysicalFunction struct {
+	Device
+	MaxVFNum int               `json:"max_vf_num,omitempty"`
+	VFs      []VirtualFunction `json:"vfs,omitempty"`
+}
+
+type VirtualFunction struct {
+	Device
+	ID int `json:"id"`
+	// Address of the (parent) Physical Function this Virtual Function pertains to.
+	ParentAddress *pciaddr.Address `json:"parent_address,omitempty"`
```

Comment:
Should we just store a pointer to the parent `pkg/pci.Device` struct instead?

Reply:
Come to think of it, could we just use a single `Function` struct to describe all of the SR-IOV functions, both physical and virtual?

```go
// Function describes an SR-IOV physical or virtual function. Physical functions
// will have no Parent Function struct pointer and will have one or more Function
// structs in the Functions field.
type Function struct {
    Device // All Functions are PCI Devices
    // Parent contains a pointer to the parent physical function.
    // Will be empty when this is a physical function
    Parent *Function `json:"parent,omitempty"`
    // MaxVFs contains the maximum number of supported virtual
    // functions for this physical function
    MaxVFs int `json:"max_vfs"
    // Functions contains the physical function's virtual functions
    Functions []*Function `json:"functions"`
}
```

You could create a simple `IsPhysical` utility method on `pkg/pci.Function`:

```go
// IsPhysical returns true if the PCIe function is a physical function, false
// if it is a virtual function
func (f *Function) IsPhysical() bool {
    return f.Parent == nil
}
```

Reply:
Interesting angle, I don't have again strong feelings about this idea. With respect the current approach, we lose something, we gain something. I'll file a separate PR so we can compare the two approaches, and I'll close the one we like less (with rationale of course).